### PR TITLE
Entropy coding option

### PR DESCRIPTION
**Pull request by deiteris**
In h264 codec, CABAC entropy coding may save up to 20% of the bitrate budget, but increases the decoder latency by 2+ times depending on the target bitrate and may also increase encoder load. This significantly reduces the decoder throughput which may be incomparable to gained quality, so CAVLC can be used for very high bitrate or high encoder/decoder load scenarios without significant perceived quality loss.

Graph with CAVLC
image

Graph with CABAC
image

### CODE REVIEW

**Comment by zarik5** [comment]
I think the default value should be CAVLC, since it's the option that behaves more consistently and gives overall lower latency.

File: alvr/server/cpp/platform/linux/EncodePipelineNvEnc.cpp [hunk 1]
```diff
@@ -51,6 +51,14 @@ alvr::EncodePipelineNvEnc::EncodePipelineNvEnc(VkFrame &input_frame,
     case ALVR_CODEC_H264:
         AVUTIL.av_opt_set(encoder_ctx, "preset", "llhq", 0);
         AVUTIL.av_opt_set(encoder_ctx, "zerolatency", "1", 0);
+
+        switch (settings.m_entropyCoding) {
+        case ALVR_CABAC:
+            AVUTIL.av_opt_set(encoder_ctx, "coder", "ac", 0);
```
Comment (nowrep) [inline alvr/server/cpp/platform/linux/EncodePipelineNvEnc.cpp:57]: First, rebase the PR on master.

It should be `AVUTIL.av_opt_set(encoder_ctx->priv_data, ...`

And are you sure `ac` `vlc` is correct?
https://github.com/FFmpeg/FFmpeg/blob/eef763c7057a7f5f4b7dae7855d07b2a6da8b537/libavcodec/vaapi_encode_h264.c#L1237

Comment (deiteris) [reply alvr/server/cpp/platform/linux/EncodePipelineNvEnc.cpp:57]: Yes, you can find the description of those options in FFmpeg Codecs documentation under VAAPI encoders. `ac` and `cabac` correspond to CABAC and `vlc` and `cavlc` correspond to CAVLC.
https://ffmpeg.org/ffmpeg-codecs.html#VAAPI-encoders

Comment (deiteris) [reply alvr/server/cpp/platform/linux/EncodePipelineNvEnc.cpp]: Should ` AVUTIL.av_opt_set(encoder_ctx->priv_data, ...` be used this way in both nvenc and vaapi code?

Comment (nowrep) [reply alvr/server/cpp/platform/linux/EncodePipelineNvEnc.cpp]: I think so, yes.

**Comment by deiteris** [comment]
Ready to merge if it's all good. Though would be great to see some testing on nvidia cards and Linux.

**Review by nowrep — APPROVED** [review]
VAAPI part tested and it does actually reduce decoder latency.

**Pull request merged**